### PR TITLE
Warmup submission for different game modes

### PR DESCRIPTION
**Pull request by vunyunt**
One thing to note is that the `legacyGamemodeId` map is defined in `types.ts`, not 100% sure if it fits there

### CODE REVIEW

File: server/api.ts [hunk 1]
```diff
@@ -103,9 +103,17 @@ const parseWarmup = async (warmup: string, mod: WarmupMod, tourney: string) => {
     throw new Error("This beatmap is part of the map pool");
   }
 
+  // Fetch tourney object to get the correct game mode
+  const tourneyData = await Tournament.findOne({ code: tourney }).orFail();
+
+  // Make taiko the default mode (old tourney data doesn't seem to have mode defined)
+  tourneyData.mode ??= "taiko";
```
Comment (cory2067) [inline server/api.ts:110]: didn't know this operator existed!!
cool

File: server/types.ts [hunk 2]
```diff
@@ -29,4 +29,15 @@ export type DiscordAccount = {
 
 type GameMode = "taiko" | "catch"; // supported osu! gamemodes
 
+/**
+ * Maps gamemodes to their legacy ID
+ */
+export const legacyGamemodeId = Object.freeze({
```
Comment (cory2067) [inline server/types.ts:35]: a function for this is also defined in util.ts, getGamemodeId -- we should probably consolidate it to one place

**Review by cory2067 — APPROVED** [review]

File: server/api.ts [hunk 3]
```diff
@@ -124,7 +132,10 @@ const parseWarmup = async (warmup: string, mod: WarmupMod, tourney: string) => {
     throw new Error("Warmup map too long (max 3 minutes)");
   }
 
-  return `https://osu.ppy.sh/beatmapsets/${mapData.beatmapSetId}#taiko/${warmupMapId}`;
+  return `https://osu.ppy.sh/beatmapsets/${mapData.beatmapSetId}#${
+    // GameMode defines catch but in beatmap url it's fruits
+    tourneyData.mode == "catch" ? "fruits" : tourneyData.mode
```
Comment (ThePooN) [inline server/api.ts:137]:
```suggestion
    tourneyData.mode === "catch" ? "fruits" : tourneyData.mode
```
for consistency (with the rest of this file, at least)

Comment (ThePooN) [inline server/api.ts:137]: hardcoding this with only 2 options is shooting ourselves in the foot (again)

**Review by ThePooN — COMMENTED** [review]
hello there

File: server/api.ts (hunk 3 above)
Comment (vunyunt) [reply server/api.ts:137]: This should work for all game modes (as long as `tourneyData.mode` is defined properly for `osu`/`mania` if these game modes exists in future tournies). It only converts `catch` to `fruits` and passes everything else in as is.

File: server/types.ts (hunk 2 above)
Comment (vunyunt) [reply server/types.ts]: Ahh didn't know that, d4c25040276da79ecc5d20a5ff4b2d3a4cbaa008 now uses a function defined in `utils.ts`

I also went ahead and added other known game modes to `GameMode`'s typedef and the `MODE_TO_ID` map. In both of these both `catch` and `fruits` are defined though, not sure if that's desirable or `GameMode` typedef

File: server/api.ts (hunk 3 above)
Comment (vunyunt) [reply server/api.ts]: Applied in d4c25040276da79ecc5d20a5ff4b2d3a4cbaa008, though it's moved to `utils.ts`

**Pull request merged**